### PR TITLE
many fixes

### PR DESCRIPTION
**Pull request by pjordaan**
fixes for packages that do define and calling the factory directly instead (for example zenscroll).
```
define(['dependencies'], factory()});
```

wrong module name for external modules with 'src' somewhere in the path. (for example angular2-swiper/src/ks-swiper.module)

Possible XSS on module name.

Do not automatically remove .<something> from a module name, it's not always '.ts'

Enable ts-helpers to reuse ts transpile helper functions.

Enable emitting deocrator metadata for Angular DI.

Added optional config for FileConfig to provide non-static options.

Ignore comments with require() statements.

Make angular templates always relative and let the import provide the full path instead.

Parse less to css for angular templates possiblity.

Ignore .d.ts entirely. This is because some imports fail because end up at the definition file, while a source file is available too. For example @angular/platform-browser/animations resolved to node_modules/@angular/platform-browser/animations/index.d.ts, while it should resolve to node_modules/@angular/platform-browser/bundles/platform-browser-animations.umd.js by reading the package.json.

### CODE REVIEW

File: src/Bundler/Processor/TsContentProcessor.php [hunk 1]
```diff
@@ -35,8 +35,8 @@ public function transpile(string $cwd, ContentItem $item): void
     {
         $module_name = $item->module_name;
 
-        if (false !== ($i = strrpos($module_name, '.'))) {
-            $module_name = substr($module_name, 0, $i);
+        if (preg_match('/\.ts$/i', $module_name)) {
+            $module_name = preg_replace('/\.ts$/i', '', $module_name);
```
Comment (yannickl88) [inline src/Bundler/Processor/TsContentProcessor.php]: can you not always replace it? because `preg_replace` does not do anything if it does not match anything.

File: src/EventListener/AngularHtmlListener.php [hunk 2]
```diff
@@ -51,7 +51,7 @@ public function onPostTranspile(AssetEvent $event): void
         $content = preg_replace_callback(
             '/templateUrl\s*:(\s*[\'"`](.*?)[\'"`]\s*)/m',
             function ($match) use ($file, $reader) {
-                return 'template: ' . json_encode($this->getCompiledAssetFor($match[2], $file, $reader));
+                return 'template: ' . json_encode($this->getCompiledAssetFor('./' . $match[2], $file, $reader));
```
Comment (yannickl88) [inline src/EventListener/AngularHtmlListener.php]: what if these are absolute paths? Or is that not possible?

File: src/Config/ConfigInterface.php [hunk 3]
```diff
@@ -61,7 +61,6 @@ public function getPlugins(): array;
     /**
      * Return the output folder in which to dump the compiled assets.
      *
-     * @param bool $include_public_folder Whether to include the web/ directory
```
Comment (nicoschoenmaker) [inline src/Config/ConfigInterface.php]: Not sure why this doc is removed?

File: src/Bundler/Runner/js/tsc.js [hunk 4]
```diff
@@ -5,9 +5,11 @@ function compile(source) {
         compilerOptions: {
             inlineSourceMap: false,
             skipLibCheck: true,
+            importHelpers: true,
             target: ts.ScriptTarget.ES5,
             module: ts.ModuleKind.CommonJS,
             moduleResolution: ts.ModuleResolutionKind.NodeJs,
+            emitDecoratorMetadata: true,
```
Comment (nicoschoenmaker) [inline src/Bundler/Runner/js/tsc.js]: Why would we enable this?

The docs of typescript point to this issue: https://github.com/Microsoft/TypeScript/issues/2577

But that does not tell me anything about why this is (or isn't) a good idea.

File: src/Config/FileConfig.php [hunk 5]
```diff
@@ -40,10 +40,14 @@ public function __construct(
         array $plugins = [],
         bool $dev = false,
         EventDispatcherInterface $dispatcher = null,
-        LoggerInterface $logger = null
+        LoggerInterface $logger = null,
+        array $non_static_options = []
```
Comment (nicoschoenmaker) [inline src/Config/FileConfig.php]: This added argument is not documented in the docblock.

File: src/EventListener/AngularHtmlListener.php [hunk 6]
```diff
@@ -81,9 +81,7 @@ private function getCompiledAssetFor(string $linked_file, File $owning_file, Rea
     {
         $file_path = $linked_file;
 
-        if ($file_path[0] === '.' && $owning_file->dir !== '.' && !empty($owning_file->dir)) {
-            $file_path = $owning_file->dir . '/' . $file_path;
-        }
+        $file_path = preg_replace('/\.css$/', '.less', $owning_file->dir . '/' . $file_path);
```
Comment (nicoschoenmaker) [inline src/EventListener/AngularHtmlListener.php]: This method is called for both style and for the template, so hardcoding anything with `.css` or `.less` doesn't feel right.

File: src/Resources/require.js [hunk 7]
```diff
@@ -133,12 +133,13 @@
...
         if (_modules[name]) {
-            throw new ModuleRedeclareError(name);
+            throw new ModuleRedeclareError(name, _modules[name]._stack);
         }
 
         _modules[name] = {
             _initializer: initializer,
-            _module: null
+            _module: null,
+            _stack: (new TypeError).stack
```
Comment (nicoschoenmaker) [inline src/Resources/require.js]: What is the impact on this for client side performance? I understand it's nice for debugging, but wouldn't this mean creating >100 objects for a javascript intensive app?

File: test/Config/FileConfigTest.php [hunk 8]
```diff
@@ -28,7 +28,6 @@ public function testMinimal()
         self::assertSame([], $config->getEntryPoints());
         self::assertSame([], $config->getAssetFiles());
         self::assertSame(self::EXPECTED_OUTPUT, $config->getOutputFolder());
-        self::assertSame('dev', $config->getOutputFolder(false));
```
Comment (nicoschoenmaker) [inline test/Config/FileConfigTest.php]: Why are these test-cases removed?

File: src/Bundler/Runner/js/tsc.js (hunk 4 above)
Comment (pjordaan) [reply src/Bundler/Runner/js/tsc.js]: Angular require this to be turned on.

File: src/Config/ConfigInterface.php (hunk 3 above)
Comment (pjordaan) [reply src/Config/ConfigInterface.php]: merge mistake....

File: src/EventListener/AngularHtmlListener.php (hunk 2 above)
Comment (pjordaan) [reply src/EventListener/AngularHtmlListener.php]: technically it's possible, but in general it doesn't happen, but I can add it if you want....

File: src/Bundler/Pipeline/ContentPipeline.php [hunk 9]
```diff
@@ -94,7 +94,7 @@ public function push(array $dependencies, ReaderInterface $file_reader, File $ta
             $module_name = $file->getName();
 
             if (!empty($this->config->getSourceRoot())
-                && false !== strpos($module_name, $this->config->getSourceRoot())
+                && 0 === strpos($module_name, $this->config->getSourceRoot())
```
Comment (nicoschoenmaker) [inline src/Bundler/Pipeline/ContentPipeline.php]: This is a fix for probably a bug that exists for which no unit-test has been added.

File: src/Bundler/Processor/ModuleProcessor.php [hunk 10]
```diff
@@ -26,7 +26,9 @@ public function peek(string $cwd, ContentState $state): void
 
     public function transpile(string $cwd, ContentItem $item): void
     {
-        $js  = "register('" . $item->module_name . "', function (define, require, module, exports) {\n";
+        $js  = "register("
+            . json_encode($item->module_name, JSON_UNESCAPED_SLASHES)
```
Comment (nicoschoenmaker) [inline src/Bundler/Processor/ModuleProcessor.php]: This is a fix for probably a bug that exists for which no unit-test has been added.

File: src/Config/FileConfig.php [hunk 11]
```diff
@@ -34,16 +34,21 @@
...
+     * @param mixed[] $non_static_options
      */
     public function __construct(
         string $config_file,
         array $plugins = [],
         bool $dev = false,
         EventDispatcherInterface $dispatcher = null,
-        LoggerInterface $logger = null
+        LoggerInterface $logger = null,
+        array $non_static_options = []
```
Comment (nicoschoenmaker) [inline src/Config/FileConfig.php]: What is the use case for these non static options? Especially the docblock type `mixed` scares me.

Maybe we should remove the `FileConfig` and move the `ArrayConfig` from the bundle here instead. Or make the `FileConfig` class a wrapper around the `ArrayConfig` class.

File: src/Plugin/LessPlugin.php [hunk 12]
```diff
@@ -25,6 +27,15 @@ public function activate(PluginApi $plugin_api): void
         if ($plugin_api->getConfig()->isDev()) {
             $less_collector = new CachedImportCollector($less_collector, $plugin_api->getCache());
         }
+        $plugin_api->getConfig()->getEventDispatcher()->addListener(
+            AssetEvents::FIND_SOURCE,
+            function (FindSourceEvent $ev) {
+                $file_path = preg_replace('/\.css$/', '.less', $ev->getFilepath());
```
Comment (nicoschoenmaker) [inline src/Plugin/LessPlugin.php]: Why wouldn't the Angular component refer to the `.less` file immediately? That would also help the IDE to find the file.

File: src/Import/BuiltIn/AngularImportCollector.php [hunk 13]
```diff
@@ -6,16 +6,25 @@
...
+use Hostnet\Component\Resolver\Plugin\PluginApi;
 
 /**
  * Angular asset resolver.
  */
 final class AngularImportCollector implements ImportCollectorInterface
 {
+    private $plugin_api;
+
+    public function __construct(PluginApi $plugin_api)
```
Comment (nicoschoenmaker) [inline src/Import/BuiltIn/AngularImportCollector.php]: This appears to only need the `EventDispatcher`.

File: src/Plugin/LessPlugin.php (hunk 12 above)
Comment (pjordaan) [reply src/Plugin/LessPlugin.php]: if Angular Ahead of Time would be used, using .less inmediately would load the less file in as css.

File: src/Bundler/Processor/ModuleProcessor.php (hunk 10 above)
Comment (pjordaan) [reply src/Bundler/Processor/ModuleProcessor.php]: this is just a common XSS fix

File: src/Config/FileConfig.php (hunk 11 above)
Comment (pjordaan) [reply src/Config/FileConfig.php]: I had the issue I have one setting which is not statically determined. I can generate a json file and use this class or indeed make a wrapper class for just one setting or I could just add this argument to keep it simple

Comment (yannickl88) [reply src/Config/FileConfig.php]: seems to be only used in the test? There is an interface, just mock it?

File: test/Config/FileConfigTest.php [hunk 14]
```diff
@@ -68,7 +68,13 @@ public function testIncludePaths()
...
     public function testMinimalRelativeNodePaths()
     {
-        $config = new FileConfig(__DIR__.'/../fixtures/configs/minimal-relative.config.json', [], true);
+        $config = new FileConfig(
+            __DIR__.'/../fixtures/configs/minimal-relative.config.json',
+            [],
+            true,
+            null,
+            null
+        );
```
Comment (yannickl88) [inline test/Config/FileConfigTest.php]: is this change still needed? I think those extra arguments are defaults

File: src/EventListener/AngularHtmlListener.php [hunk 15]
```diff
@@ -79,12 +81,16 @@ function ($match) use ($file, $reader) {
      */
     private function getCompiledAssetFor(string $linked_file, File $owning_file, ReaderInterface $reader): string
     {
-        $file_path = $linked_file;
-
-        if ($file_path[0] === '.' && $owning_file->dir !== '.' && !empty($owning_file->dir)) {
-            $file_path = $owning_file->dir . '/' . $file_path;
+        if (substr($linked_file, 1, 0) === '/') {
```
Comment (nicoschoenmaker) [inline src/EventListener/AngularHtmlListener.php]:
```bash
php -r "var_dump(substr('ABC', 1, 0));"
```
```
string(0) ""
```
Based on that, this if can not become true. There should probably be a unit-test for this case.

**Comment by nicoschoenmaker** [comment]
Can we tear this pull request in 10 pieces or so?

This appears to fix a lot of issues at the same time, it's hard to see how these are related now. Or whether all the unit-tests are complete, etc.
